adapters/adhese/adhese.go

```diff
@@ -0,0 +1,298 @@
+package adhese
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type AdheseAdapter struct {
+	http             *adapters.HTTPAdapter
+	endpointTemplate template.Template
+	dummyCacheBuster int
+}
+
+type AdheseKeywordsParams struct {
+	Key    string   `json:"key,omitempty"`
+	Values []string `json:"value,omitempty"`
+}
+
+func (a *AdheseAdapter) Name() string {
+	return "adhese"
+}
+
+func (a *AdheseAdapter) SkipNoCookies() bool {
+	return false
+}
+
+func extractSlotParameter(parameters openrtb_ext.ExtImpAdhese) string {
+	return fmt.Sprintf("/sl%s-%s", parameters.Location, parameters.Format)
+}
+
+func extractTargetParameters(parameters openrtb_ext.ExtImpAdhese) string {
+	if parameters.Keywords == nil || len(parameters.Keywords) == 0 {
+		return ""
+	}
+	m := make(map[string][]string)
+	for _, kv := range parameters.Keywords {
+		for _, tv := range kv.Values {
+			cur, _ := m[kv.Key]
+			new := cur[:]
+			m[kv.Key] = append(new, tv)
+		}
+	}
+
+	var parametersAsString = ""
+
+	for k, v := range m {
+		parametersAsString += "/" + k + strings.Join(v, ";")
+	}
+	params := strings.Split(parametersAsString, "/")
+	sort.Strings(params)
+	return strings.Join(params, "/")
+
+}
+
+func extractGdprParameter() string {
+	//const gdprParams = (gdprConsent && gdprConsent.consentString) ? [ 'xt' + gdprConsent.consentString, 'tlall' ] : [];
+	return ""
+}
+
+func extractRefererParameter() string {
+	//const refererParams = (refererInfo && refererInfo.referer) ? [ 'xf' + base64urlEncode(refererInfo.referer) ] : [];
+	return ""
+}
+
+func (a *AdheseAdapter) generateCacheBuster() string {
```

Comment:
I imagine we don't need that, do we? The request is executed server side so no caches involved